controllers: Deployment with no available pods can be Available, which is confusing

smarterclayton opened the pull request:
A deployment with no available replicas can be Available. This is a consequence of an implementation choice and has been consistently represented for most of the lifecycle of the v1 apps group, and is arguably the user choosing to allow the minimum available pods to be 0 by default.  We should clarify in API docs and our available message when this occurs to alert the consumer to the potential error.

Also, the Recreate strategy reports Available inconsistently with the definition, resulting in Availability = False during upgrades despite having an effective minimum availability of zero.

/kind bug

```release-note
The deployment Available condition reports true there are at least as many available pods as the minimum availability value.  Since the default deployment has a minimum availability of zero (rolling upgrade strategy with maxSurge 0 and maxUnavailable defaulting to 1), users may be surprised to see the Available condition report true during rolling upgrades when there are no pods.  To ensure the behavior is clear, clarify the API documentation for the conditions field on deployments with details of the default condition types and return a distinct message on the Available condition when zero available replicas are found.
```

```docs

```

alculquicondor commented on pkg/controller/deployment/deployment_controller_test.go:
```diff
@@ -150,28 +159,43 @@ type fixture struct {
-	actions []core.Action
-	objects []runtime.Object
+	actions     []core.Action
+	objectTests []func(runtime.Object)
```
it looks like this is all about comparing the object.

Maybe the fixture can simply have a boolean to validate the equality of objects. You can use `cmpopts` to ignore certain fields.

smarterclayton replied:
I also need to add tests verifying correct fenceposts in a variety of surge conditions

aojea commented on pkg/controller/deployment/deployment_controller_test.go:
```diff
@@ -150,28 +159,43 @@ type fixture struct {
+	f.objectTests = append(f.objectTests, nil)
+}
+
+func (f *fixture) expectUpdateDeploymentActionFunc(d *apps.Deployment, fn func(runtime.Object)) {
```
linter complains this is `unused`

atiratree commented:
This has been discussed in Sig Apps meeting on Oct 17. The current behaviour is unintentional, but we should keep it in order not to break the v1 API. This could potentially break users with existing clusters, since this has implications for anybody observing this condition. Eg. monitoring correct workoads that expect 0 replicas in rollouts could suddenly start triggering notifications. Nevertheless, I agree the proposed change could be beneficial to most of the users so they would be notified that their workload has 0 replicas.

But in order not to break the current users that are for better or worse depending on this behaviour, it might be better to warn when such attempted rollouts occur:
- we could create an event every time a new rollout occurs whose constraints could result in 0 replicas (similar to cron jobs PR https://github.com/kubernetes/kubernetes/pull/106455/files)
- generate a warning on the CLI side when a deployment is created/updated
- generate a warning event on the API or controller side when an update to a deployment occurs

this could be any combination of these 3 options

smarterclayton commented:
>  but we should keep it in order not to break the v1 API. This could potentially break users with existing clusters, since this has implications for anybody observing this condition.

I'm gonna need more detail than that, many bugs that involve observing these failures incorrectly could be broken - that is weighed against all of the people who *expect* the behavior to happen, code to it, and then don't observe it.  What's the actual reason that someone would be broken?  Please describe the scenario in more detail.

smarterclayton commented:
Given the documented behavior and looking at the history of the code, I think it's unfortunate but necessary to leave this behavior in place.  But we can do better in explaining it.  Updated the PR to:

* clarify the API doc
* return a specific message when a deployment is available but with zero replicas
* add unit tests that verify condition output in more scenarios around this to avoid regressions in the future

smarterclayton commented:
Ugh, and now I realize that Recreate and Rolling are inconsistent (Recreate is available=false at 0 available replicas).  I need to reflect more but that seems wrong.

smarterclayton committed:
controllers: Deployment with no available pods can be Available (1/2)

A deployment is Available when the minimum availability of replicas
has been reached, which may include zero pods for deployments that
specify the Recreate upgrade strategy, have minUnavailable set to a
value greater than or equal to the current requested replicas, or
have a single replica without specifying a maxSurge.

Since this has been consistently reported for the majority of the
Deployment lifespan, rather than change the behavior to mitigate
the confusion we will clarify it in our API docs, in the message
returned with the Available condition, and in our user facing
documentation.

Fix a bug where a Recreate deployment is listed as not Available
during upgrade, even when it should be listed as Available because
the minimum availability of a recreate deployment is zero.

Add a unit test that verifies the expected behavior.

smarterclayton committed:
controllers: Deployment with no available pods can be Available (2/2)

A deployment is Available when the minimum availability of replicas
has been reached, which may include zero pods for deployments that
specify the Recreate upgrade strategy, have minUnavailable set to a
value greater than or equal to the current requested replicas, or
have a single replica without specifying a maxSurge.

Since this has been consistently reported for the majority of the
Deployment lifespan, rather than change the behavior to mitigate
the confusion we will clarify it in our API docs, in the message
returned with the Available condition, and in our user facing
documentation.

Update the API description of the conditions field on deployments.

atiratree commented:
The message in Available condition seems like a good compromise to let users know that their deployment might be in undesirable state.

to summarize the events approach:
- pro: user would notice even when he/she missed the possible short dip in replicas when rollout is occurring.
- contra: is that it would be quite spammy for users that do not care about this.

> A user who cares about using available to know whether they have any pods today has to define a min floor > 1

yes, at least there is a workaround for users that requireone replica to be always up

> Something as simple as explaining the fencepost condition in the Available=true message, in the API doc, and in the formal docs may be sufficient.

+1, do you want to make a PR against the formal docs, or should I start it?

atiratree commented on staging/src/k8s.io/api/apps/v1/types.go:
```diff
@@ -476,7 +476,18 @@ type DeploymentStatus struct {
+	// Available is true when the number of available replicas under the deployment's replica sets is
+	// greater than the user's defined minimum availability. The minimum availability is controlled by the
+	// update strategy - the Recreate strategy has a minimum availability of zero (always Available), and
+	// the Rolling strategy defaults to a minimum availability of zero unless replicas is greater than 1,
```
> // Absolute number is calculated from percentage by rounding down.
> // Defaults to 25%.
> MaxUnavailable

it defaults to 25% `maxUnavailable` and 25% `maxSurge` so in case replicas is 0-3 the minimum availability will be 100%

smarterclayton replied:
Coming back to this comment, now I'm confused (which doesn't bode well).  I think you're saying "the fencepost protection for 25% ensures that if `total pods * 25% < 1, your minimum availability is 1"?

atiratree replied:
I see, it depends how we want to convey the message. If we define cases with zero minimum availability or cases with positive/general minimum availability. 

We mention `minimum availability` and jump to explaining zero minimum availability. I think it might be useful to at least explain the term first, mainly for Rolling strategy (eg. just by mentioning `.spec.replicas - maxUnavailable` or `.spec.replicas - .spec.strategy.rollingUpdate.maxUnavailable`). Then we could define what zero minimum availability can happen.

atiratree commented on staging/src/k8s.io/api/apps/v1/types.go:
```diff
@@ -476,7 +476,18 @@ type DeploymentStatus struct {
+	// greater than the user's defined minimum availability. The minimum availability is controlled by the
+	// update strategy - the Recreate strategy has a minimum availability of zero (always Available), and
+	// the Rolling strategy defaults to a minimum availability of zero unless replicas is greater than 1,
+	// the maxSurge parameter is specified, or maxUnavailable is smaller than the number of replicas.
```
I think it is better to describe the general working of the condition before defining special cases

smarterclayton replied:
Agree

smarterclayton replied:
Added a sentence above, and tried to capture that "available condition is a lie and can't be relied on in this case".

atiratree commented on staging/src/k8s.io/api/apps/v1/types.go:
```diff
@@ -476,7 +476,18 @@ type DeploymentStatus struct {
+	// update strategy - the Recreate strategy has a minimum availability of zero (always Available), and
+	// the Rolling strategy defaults to a minimum availability of zero unless replicas is greater than 1,
+	// the maxSurge parameter is specified, or maxUnavailable is smaller than the number of replicas.
+	// Progressing is true when the deployment is trying to reach minimum availability on a new replica set.
```
it is trying to reach the full availability

atiratree replied:
and can become false when the progress times out

smarterclayton replied:
Tweaked this even more based on your comment.

atiratree commented on pkg/controller/deployment/deployment_controller_test.go:
```diff
@@ -362,6 +388,190 @@ func TestReentrantRollback(t *testing.T) {
+
+	simpleReplicaSetStatusFn := func(specReplicas, replicas, ready, available int32, fns ...modifierFunc) objectsFunc {
+		return func(t *testing.T) ([]*apps.Deployment, []*apps.ReplicaSet) {
+			d := newDeployment("foo", int(specReplicas), nil, nil, nil, map[string]string{"foo": "bar"})
```
note: by default this is set to maxUnavailable = 25% and maxUnavailable = 25% in apiserver

in this case this will be then set to maxUnavailbale = 1 in `ResolveFenceposts`

If we are testing only this case, maybe we should say so explicitely in the test description/function name

atiratree commented on pkg/controller/deployment/deployment_controller_test.go:
```diff
@@ -362,6 +388,190 @@ func TestReentrantRollback(t *testing.T) {
+			wantActionsFn: expectSimpleDeploymentStatus(false, 2, 2, 0),
+		},
+		{
+			name:          "rolling is available when all pods unready because minAvailable=0",
```
can you please elaborate more on this test?

smarterclayton commented:
> +1, do you want to make a PR against the formal docs, or should I start it?

I think so - it would be good to have both of us try to explain it, and then converge to a consistent description across both?

smarterclayton commented on staging/src/k8s.io/api/apps/v1/types.go:
```diff
@@ -493,12 +504,15 @@ type DeploymentConditionType string
 	// Available means the deployment is available, ie. at least the minimum available
-	// replicas required are up and running for at least minReadySeconds.
+	// replicas required are up and running for at least minReadySeconds. A deployment
+	// that allows a minimum availability of zero (such as with the Recreate strategy,
```
Yeah, striving for clarity here.  We're trying to indicate three things:

1. by default you aren't going to be available during upgrades (with the defaults of rolling update and replicas 1)
2. if you stay with RollingUpdate, you need to a) use max surge and not maxUnavailable, or b) set replicas 1
3. if you are Recreate, you are not going to remain available during upgrades

I may break this into two separate setups, but it's really important that people know that you are not available by default and how to get to availability.

smarterclayton commented on pkg/controller/deployment/util/deployment_util.go:
```diff
@@ -434,8 +434,16 @@ func ReplicasAnnotationsNeedUpdate(rs *apps.ReplicaSet, desiredReplicas, maxRepl
+		return 0
+	case deployment.Spec.Strategy.Type == apps.RecreateDeploymentStrategyType:
+		// the recreate strategy allows all replicas to be unavailable during upgrade
+		return *(deployment.Spec.Replicas)
```
We have historically reserved the right to say that consistency of client expectation overrides some specific behaviors - generally if the implemented feature does not accomplish the goal it intended to serve due to bugs.  In this case, the recreate deployment is clearly Available=true during upgrade, and someone depending on Available=false some of the time (not deployment) isn't going to get 100% behavior, so the field is only partially useful, and the user should be doing something more accurate.

We should probably codify that in a way that allows reviewers to reach consensus and preserve the correct default bias not to break behavior while still having a set of guidelines to deal with the cases where that results in a worse system (they should be rare, but they exist).

smarterclayton committed:
wip comments

dims commented:
/milestone v1.27

soltysh commented:
/assign @atiratree

sowmyav27 commented:
@smarterclayton @atiratree Hi, I am part of Bug Triage Shadow team for k8s 1.27 release. I wanted to confirm this PR is on track for 1.27?

atiratree commented:
I will try to revisit this and pick the documentation part

sowmyav27 commented:
Hi @atiratree I am part of Bug Triage Shadow team for k8s 1.27 release. Can you please confirm if this is on track for 1.27 Release ?

smarterclayton commented:
Will refresh this shortly and work with @atiratree

atiratree commented:
@smarterclayton here is the proposal for the explanation: https://github.com/kubernetes/website/pull/39892

dims commented:
Which one goes in first?
- https://github.com/kubernetes/website/pull/39892 
- https://github.com/kubernetes/kubernetes/pull/113228

atiratree commented:
discussion about this in the website PR: https://github.com/kubernetes/website/pull/39892#issuecomment-1465999145

smarterclayton commented on staging/src/k8s.io/api/apps/v1/types.go:
```diff
@@ -493,12 +504,15 @@ type DeploymentConditionType string
-	// considered when a new replica set is created or adopted, and when new pods scale
-	// up or old pods scale down. Progress is not estimated for paused deployments or
-	// when progressDeadlineSeconds is not specified.
+	// considered when a new replica set is created or adopted. Progress is not
```
See my new wording and the tweak to progressDeadlineSeconds.

atiratree replied:
+1

smarterclayton commented:
Pushed a commit that has just the potential wording changes (given that's the most confusing part) for review.

smarterclayton commented:
Going to wait on this until 1.28

atiratree commented on staging/src/k8s.io/api/apps/v1/types.go:
```diff
@@ -520,12 +536,18 @@ type DeploymentConditionType string
+	// zero when using the Recreate strategy, or when setting maxUnavailable in
+	// RollingUpdate deployments greater than or equal to the number of replicas, or
+	// when a RollingUpdate deployment has only one replica and any non-zero
+	// maxUnavailable value.
```
Is it useful to have 2 definitions? Don't we want to make only one definition of Available and then copy it to the Conditions doc?

smarterclayton replied:
Since few people reference the constant out of tree, i think we could have the authoritative definition on Conditions and have a “see type DeploymentSpecConditions” for all the constants.

soltysh commented:
> Going to wait on this until 1.28

Are you also planning to backport that to previous release, or given the changes in behavior - no?

dims commented:
@atiratree poke!

atiratree commented:
did you mean to poke @smarterclayton instead?

smarterclayton commented:
No backport.  Will revisit this week with latest review

dims commented:
@smarterclayton  needs rebase!

moficodes commented:
@atiratree @smarterclayton 

1.28 Bug Triage Shadow here.

Just checking-in, if this is still on track for the 1.28 release?

furkatgofurov7 commented:
> @atiratree @smarterclayton
> 
> 1.28 Bug Triage Shadow here.
> 
> Just checking-in, if this is still on track for the 1.28 release?

@smarterclayton @liggitt  Hey folks!

Reiterating on above comment and FYI the code freeze is starting [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496) (about 3 weeks from now). 

Since this PR is targeting the 1.28 milestone, we would like to know if this still planned for this release?

moficodes commented:
Week 9 update: Code Freeze is approaching!

Hey! We are currently in Week 9, and Code Freeze is just over a week away. It will begin at 01:00 UTC on Wednesday, 19th July 2023 / 18:00 PDT on Tuesday, 18th July 2023. Please ensure all necessary submissions are made before the deadline. Let me know if you need any assistance. :)
